AVRO-3856: [C#] Hardcode MaxDepth property to 192 when using Newtonsoft JsonReader to load Avro schema up to 64 depth levels

tradercentric opened the pull request:
<!--

*Thank you very much for contributing to Apache Avro - we are happy that you want to help us improve Avro. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Avro a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/AVRO/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "AVRO-XXXX: [component] Title of the pull request", where *AVRO-XXXX* should be replaced by the actual issue number. 
    The *component* is optional, but can help identify the correct reviewers faster: either the language ("java", "python") or subsystem such as "build" or "doc" are good candidates.  

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests. You can [build the entire project](https://github.com/apache/avro/blob/master/BUILD.md) or just the [language-specific SDK](https://avro.apache.org/project/how-to-contribute/#unit-tests).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Every commit message references Jira issues in their subject lines. In addition, commits follow the guidelines from [How to write a good git commit message](https://chris.beams.io/posts/git-commit/)
    1. Subject is separated from body by a blank line
    1. Subject is limited to 50 characters (not including Jira issue reference)
    1. Subject does not end with a period
    1. Subject uses the imperative mood ("add", not "adding")
    1. Body wraps at 72 characters
    1. Body explains "what" and "why", not "how"

-->

## What is the purpose of the change

I am an end-user/developer to use Confluent Kafka to consume a Avro message where the number of nested structures is around 20.   The producer/vendor cannot reduce the depth level of the Avro schema.    Confluent support has assisted me to open an issue with Apache Avro project.  The issue is logged in https://issues.apache.org/jira/browse/AVRO-3856.    I am in need of code fix so I can continue.

The current code in Schema.cs parsing Avro schema using JObject.Parse(json), JArray.Parse(json) do not have option to override max depth level.  On https://github.com/JamesNK/Newtonsoft.Json/pull/2904, Newtonsoft Author advised to use JObject.Load(Jsonreader), so we can set the MaxDepth in JsonReader once instantiated.   

In doing so, I also noticed the Avro schema depth level is not counted the same as in JsonReader's.  The default depth level limit of 64 in JsonReader can only support around 20 levels of Avro schema.

When putting break point in JsonReader Push method (called by Load method) with various Avro schema depth levels, here is my observation:

Avro Schema Depth | JsonReader Depth Count
4 | 11
16 | 44
32 | 92
64 | 188

To compensate, I had hardcoded JsonReader MaxDepth to 192 to support parsing of Avro Schema with 64 level depth.

## Verifying this change

This change added tests and can be verified as follows:

I added 3 test cases in SchemaTests.cs:

Parse16DepthLevelSchemaTest
Parse32DepthLevelSchemaTest
Parse64DepthLevelSchemaTest


## Documentation

- Does this pull request introduce a new feature? no
- If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

tradercentric committed:
Compensate Newtonsoft JsonReader over-countered Avro schema depth by 3x by setting JsonReader MaxDepth to 192.  Also I took the Newtonsoft author advice to use JObject.Load(JsonReader) and JArray.Load(JsonReader) instead of their Parse method.  So I can set the JsonReader MaxDepth.

tradercentric committed:
Add comment to refer to the Avro open issue.

KalleOlaviNiemitalo reviewed (changes requested):
Does Avro.IO.JsonDecoder also need to be changed?  I mean, if Avro supports a 64-level record schema, then it should also be possible to take construct an instance of such a record, encode it to JSON, and decode it back.

tradercentric committed:
Update the comment first so make it clear it is not a bug in Newtonsoft JsonReader

tradercentric commented on lang/csharp/src/apache/main/Schema/Schema.cs:
```diff
@@ -243,11 +244,28 @@ internal static Schema Parse(string json, SchemaNames names, string encspace)
                 bool IsArray = json.StartsWith("[", StringComparison.Ordinal)
                     && json.EndsWith("]", StringComparison.Ordinal);
-                JContainer j = IsArray ? (JContainer)JArray.Parse(json) : (JContainer)JObject.Parse(json);
+                JContainer j = IsArray ? (JContainer)JArray.Load(reader) : (JContainer)JObject.Load(reader);
```
Ack.  In progress to look into it

tradercentric replied:
I did not add the tests because I can follow exactly what JObject.Parse has done after JObject.Load:

try
{
    bool IsArray = json.StartsWith("[", StringComparison.Ordinal)
        && json.EndsWith("]", StringComparison.Ordinal);
    JContainer j = IsArray ? (JContainer)JArray.Load(reader) : (JContainer)JObject.Load(reader);

    // When replacing JArray.Parse and JObject.Parse with JArray.Load and JObject.Load,
    // we will need this check following what Newtonsoft.Json does.
    while (reader.Read())
    {
        // Any content encountered here other than a comment will throw in the reader.
    }
:
:

tradercentric replied:
Please help to review.

tradercentric commented on lang/csharp/src/apache/main/Schema/Schema.cs:
```diff
@@ -243,11 +244,28 @@ internal static Schema Parse(string json, SchemaNames names, string encspace)
+            // Refer to https://github.com/JamesNK/Newtonsoft.Json/pull/2904
+            // Newtonsoft author advised to use JObject.Load/JArray.Load instead of JObject.Parse()/JArray.Parse()
+            // The reason is we can set the MaxDepth property on the JsonReader.
+            JsonReader reader = new JsonTextReader(new StringReader(json));
```
Ack.  In progress to look into it

tradercentric commented on lang/csharp/src/apache/main/Schema/Schema.cs:
```diff
@@ -243,11 +244,28 @@ internal static Schema Parse(string json, SchemaNames names, string encspace)
+            // 64	                188
+            // So, roughly speaking, the depth level count is about 2.75 times of Avro schema depth.
+            // Below is the hard-coded value to compensate over-counting of depth level in Newtonsoft
+            // to support Avro schema depth level to 64 slightly beyond.
```
I reworded the comment first and committed.

tradercentric commented on lang/csharp/src/apache/main/Schema/Schema.cs:
```diff
@@ -243,11 +244,28 @@ internal static Schema Parse(string json, SchemaNames names, string encspace)
+            // So, roughly speaking, the depth level count is about 2.75 times of Avro schema depth.
+            // Below is the hard-coded value to compensate over-counting of depth level in Newtonsoft
+            // to support Avro schema depth level to 64 slightly beyond.
+            reader.MaxDepth = 192;
```
I do not think I am qualified to make those changes to allow application developer to customize the limit.  I would like an enhancement.   I did more analysis on the real world Avro schema, and found out it is just 1 above the 64 max depth level defaulted in JsonReader.

tradercentric replied:
Please help to review.

tradercentric committed:
1. Added using resource
2. Followed JObject.Parse and JArray.Parse on reading the remaining json for error
3. Added a real world testcase of a schema where parsing failed.  The Avro schema is around JsonReader defaulted 64 max depth level

tradercentric commented:
Added comment of a workaround on end user/developer side in https://issues.apache.org/jira/browse/AVRO-3856.

emasab commented:
Thanks @tradercentric @KalleOlaviNiemitalo I've updated the issue so it corresponds to the fix.

tradercentric commented:
I noticed there is no deserialization error in Java Avro library with the same schema.
May I ask whose Confluent developers qualified  in improving Apache.Avro C# library can lead the effort to allow customization of max depth please?

tradercentric commented:
Hi @KalleOlaviNiemitalo , I have reworded the title to reflect the wish to hardcode value of 192 to support parsing of Avro schema up to 64 depth levels.

chetanmsft reviewed (approved):
Would be useful to have this change checked in, thanks!

chetanmsft commented:
@JamesNK.Newtonsoft.Json @opwvhk could you please take a look?

tradercentric commented:
I took the early retirement deal from my current company when I reported this issue.  Thanks for take care of it before my departure at the end of October. :)

Check this WIP pet project out:  https://strategic-retirement-397310840166.us-central1.run.app/
